python/common/org/python/types/List.java

```diff
@@ -167,8 +168,14 @@ public List(org.python.Object [] args, java.util.Map<java.lang.String, org.pytho
             int count = Math.min(size, otherSize);
 
             for (int i = 0; i < count; i++) {
-                org.python.types.Bool b =
-                    (org.python.types.Bool) this.value.get(i).__lt__(otherList.value.get(i));
+                org.python.Object result = this.value.get(i).__lt__(otherList.value.get(i));
```

Comment:
I really should've spent more time poking at the cpython source earlier >_<

This PR's current implementation seems largely similar to `PyObject_RichCompare`, but with a few differences: (i) == instead of defaulting to False should return a Bool of whether they're the same object, similarly for !=, (ii) the operators can return non-Bool (and non-NotImplemented) values, which I'm not sure works properly now, (iii) 100 byte limit on type name in TypeError string.

I should patch and test (i) and (iii), and look into the behaviour of (ii).

As for `PyObject_RichCompareBool`, it seems to be primarily coercing the result to a Bool, with an optimization for == and != with the same object.

I agree that some implementation of `PyObject_RichCompare` should be available as a method. It seems to be needed to correctly implement other datatypes like slice, tuple, complex, [list](https://github.com/python/cpython/blob/v3.6.0/Objects/listobject.c#L2251) in cpython, and it's quite annoying to have to duplicate the logic across all of them.

cc @freakboy3742: 
1. Would it be better to still duplicate this logic in the AST code, or have it invoke the static method that is needed to implement the datatypes correctly?
2. Where would be a good place to put this static method?